diff -Naur binutils-2.24-pure/bfd/configure binutils-2.24/bfd/configure
--- binutils-2.24-pure/bfd/configure	2013-12-02 01:30:30.000000000 -0800
+++ binutils-2.24/bfd/configure	2018-05-29 21:51:19.471019600 -0700
@@ -12199,7 +12199,7 @@
 
 NO_WERROR=
 if test "${ERROR_ON_WARNING}" = yes ; then
-    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror"
+    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror -Wno-implicit-fallthrough"
     NO_WERROR="-Wno-error"
 fi
 
diff -Naur binutils-2.24-pure/bfd/warning.m4 binutils-2.24/bfd/warning.m4
--- binutils-2.24-pure/bfd/warning.m4	2013-11-04 07:33:37.000000000 -0800
+++ binutils-2.24/bfd/warning.m4	2018-05-29 21:51:19.486619600 -0700
@@ -46,7 +46,7 @@
 
 NO_WERROR=
 if test "${ERROR_ON_WARNING}" = yes ; then
-    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror"
+    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror -Wno-implicit-fallthrough"
     NO_WERROR="-Wno-error"
 fi
 		   
diff -Naur binutils-2.24-pure/binutils/configure binutils-2.24/binutils/configure
--- binutils-2.24-pure/binutils/configure	2013-11-04 07:33:37.000000000 -0800
+++ binutils-2.24/binutils/configure	2018-05-29 21:51:19.549019700 -0700
@@ -11633,7 +11633,7 @@
 
 NO_WERROR=
 if test "${ERROR_ON_WARNING}" = yes ; then
-    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror"
+    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror -Wno-implicit-fallthrough"
     NO_WERROR="-Wno-error"
 fi
 
diff -Naur binutils-2.24-pure/binutils/stabs.c binutils-2.24/binutils/stabs.c
--- binutils-2.24-pure/binutils/stabs.c	2013-11-04 07:33:37.000000000 -0800
+++ binutils-2.24/binutils/stabs.c	2018-05-29 21:48:46.916278100 -0700
@@ -2697,7 +2697,7 @@
 	      ++*pp;
 	      voffset &= 0x7fffffff;
 
-	      if (**pp == ';' || *pp == '\0')
+	      if (**pp == ';' || **pp == '\0')
 		{
 		  /* Must be g++ version 1.  */
 		  context = DEBUG_TYPE_NULL;
diff -Naur binutils-2.24-pure/config/tls.m4 binutils-2.24/config/tls.m4
--- binutils-2.24-pure/config/tls.m4	2013-11-04 07:33:37.000000000 -0800
+++ binutils-2.24/config/tls.m4	2018-05-29 21:51:19.564619700 -0700
@@ -116,7 +116,7 @@
   		 gcc_cv_use_emutls, [
     gcc_cv_use_emutls=no
     echo '__thread int a; int b; int main() { return a = b; }' > conftest.c
-    if AC_TRY_COMMAND(${CC-cc} -Werror -S -o conftest.s conftest.c 1>&AS_MESSAGE_LOG_FD); then
+    if AC_TRY_COMMAND(${CC-cc} -Werror -Wno-implicit-fallthrough -S -o conftest.s conftest.c 1>&AS_MESSAGE_LOG_FD); then
       if grep __emutls_get_address conftest.s > /dev/null; then
 	gcc_cv_use_emutls=yes
       fi
diff -Naur binutils-2.24-pure/config/warnings.m4 binutils-2.24/config/warnings.m4
--- binutils-2.24-pure/config/warnings.m4	2013-11-04 07:33:37.000000000 -0800
+++ binutils-2.24/config/warnings.m4	2018-05-29 21:51:19.564619700 -0700
@@ -97,7 +97,7 @@
 		   [enable -Werror despite compiler version]),
 [], [enable_werror_always=no])
 AS_IF([test $enable_werror_always = yes],
-      [acx_Var="$acx_Var${acx_Var:+ }-Werror"])
+      [acx_Var="$acx_Var${acx_Var:+ }-Werror -Wno-implicit-fallthrough"])
  m4_if($1, [manual],,
  [AS_VAR_PUSHDEF([acx_GCCvers], [acx_cv_prog_cc_gcc_$1_or_newer])dnl
   AC_CACHE_CHECK([whether $CC is GCC >=$1], acx_GCCvers,
@@ -111,7 +111,7 @@
    [AS_VAR_SET(acx_GCCvers, yes)],
    [AS_VAR_SET(acx_GCCvers, no)])])
  AS_IF([test AS_VAR_GET(acx_GCCvers) = yes],
-       [acx_Var="$acx_Var${acx_Var:+ }-Werror"])
+       [acx_Var="$acx_Var${acx_Var:+ }-Werror -Wno-implicit-fallthrough"])
   AS_VAR_POPDEF([acx_GCCvers])])
 m4_popdef([acx_Var])dnl
 ])# ACX_PROG_CC_WARNINGS_ARE_ERRORS
diff -Naur binutils-2.24-pure/config/weakref.m4 binutils-2.24/config/weakref.m4
--- binutils-2.24-pure/config/weakref.m4	2013-11-04 07:33:37.000000000 -0800
+++ binutils-2.24/config/weakref.m4	2018-05-29 21:51:19.564619700 -0700
@@ -4,7 +4,7 @@
   AC_CACHE_CHECK([whether the target supports weak],
 		 ac_cv_have_attribute_weak, [
   weakref_m4_saved_CFLAGS="$CFLAGS"
-  CFLAGS="$CFLAGS -Werror"
+  CFLAGS="$CFLAGS -Werror -Wno-implicit-fallthrough"
   AC_TRY_COMPILE([void __attribute__((weak)) foo(void) { }],
 		 [], ac_cv_have_attribute_weak=yes,
 		 ac_cv_have_attribute_weak=no)
diff -Naur binutils-2.24-pure/gas/configure binutils-2.24/gas/configure
--- binutils-2.24-pure/gas/configure	2013-11-04 07:33:37.000000000 -0800
+++ binutils-2.24/gas/configure	2018-05-29 21:51:19.627019900 -0700
@@ -11616,7 +11616,7 @@
 
 NO_WERROR=
 if test "${ERROR_ON_WARNING}" = yes ; then
-    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror"
+    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror -Wno-implicit-fallthrough"
     NO_WERROR="-Wno-error"
 fi
 
diff -Naur binutils-2.24-pure/gold/configure binutils-2.24/gold/configure
--- binutils-2.24-pure/gold/configure	2013-11-08 02:13:48.000000000 -0800
+++ binutils-2.24/gold/configure	2018-05-29 21:51:19.642619900 -0700
@@ -6472,7 +6472,7 @@
 
 
 save_CFLAGS="$CFLAGS"
-CFLAGS="$CFLAGS -Werror -fpic -mtls-dialect=gnu2"
+CFLAGS="$CFLAGS -Werror -Wno-implicit-fallthrough -fpic -mtls-dialect=gnu2"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -6669,7 +6669,7 @@
 
 NO_WERROR=
 if test "${ERROR_ON_WARNING}" = yes ; then
-    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror"
+    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror -Wno-implicit-fallthrough"
     NO_WERROR="-Wno-error"
 fi
 
@@ -7444,7 +7444,7 @@
 fi
 
 save_CXXFLAGS="$CXXFLAGS"
-CXXFLAGS="$CXXFLAGS -Werror -gpubnames"
+CXXFLAGS="$CXXFLAGS -Werror -Wno-implicit-fallthrough -gpubnames"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 int i;
diff -Naur binutils-2.24-pure/gold/configure.ac binutils-2.24/gold/configure.ac
--- binutils-2.24-pure/gold/configure.ac	2013-11-08 02:13:48.000000000 -0800
+++ binutils-2.24/gold/configure.ac	2018-05-29 21:51:19.658219900 -0700
@@ -395,7 +395,7 @@
 dnl assuming here that there is no compiler that groks -mtls-dialect=gnu2
 dnl but does not grok -Werror.)
 save_CFLAGS="$CFLAGS"
-CFLAGS="$CFLAGS -Werror -fpic -mtls-dialect=gnu2"
+CFLAGS="$CFLAGS -Werror -Wno-implicit-fallthrough -fpic -mtls-dialect=gnu2"
 AC_COMPILE_IFELSE([
 __thread int i;
 void foo (void)
@@ -603,7 +603,7 @@
 dnl assuming here that there is no compiler that groks -gpubnames
 dnl but does not grok -Werror.)
 save_CXXFLAGS="$CXXFLAGS"
-CXXFLAGS="$CXXFLAGS -Werror -gpubnames"
+CXXFLAGS="$CXXFLAGS -Werror -Wno-implicit-fallthrough -gpubnames"
 AC_COMPILE_IFELSE([int i;], [have_pubnames=yes], [have_pubnames=no])
 CXXFLAGS="$save_CXXFLAGS"
 AM_CONDITIONAL(HAVE_PUBNAMES, test "$have_pubnames" = "yes")
diff -Naur binutils-2.24-pure/gprof/configure binutils-2.24/gprof/configure
--- binutils-2.24-pure/gprof/configure	2013-11-04 07:33:39.000000000 -0800
+++ binutils-2.24/gprof/configure	2018-05-29 21:51:19.705020000 -0700
@@ -12045,7 +12045,7 @@
 
 NO_WERROR=
 if test "${ERROR_ON_WARNING}" = yes ; then
-    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror"
+    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror -Wno-implicit-fallthrough"
     NO_WERROR="-Wno-error"
 fi
 
diff -Naur binutils-2.24-pure/ld/configure binutils-2.24/ld/configure
--- binutils-2.24-pure/ld/configure	2013-11-04 07:33:39.000000000 -0800
+++ binutils-2.24/ld/configure	2018-05-29 21:51:19.767420100 -0700
@@ -4682,7 +4682,7 @@
 
 NO_WERROR=
 if test "${ERROR_ON_WARNING}" = yes ; then
-    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror"
+    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror -Wno-implicit-fallthrough"
     NO_WERROR="-Wno-error"
 fi
 
diff -Naur binutils-2.24-pure/opcodes/configure binutils-2.24/opcodes/configure
--- binutils-2.24-pure/opcodes/configure	2013-11-04 07:33:40.000000000 -0800
+++ binutils-2.24/opcodes/configure	2018-05-29 21:51:19.829820200 -0700
@@ -11545,7 +11545,7 @@
 
 NO_WERROR=
 if test "${ERROR_ON_WARNING}" = yes ; then
-    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror"
+    GCC_WARN_CFLAGS="$GCC_WARN_CFLAGS -Werror -Wno-implicit-fallthrough"
     NO_WERROR="-Wno-error"
 fi
 
