diff -Naur gcc-4.7.4-pure/config/warnings.m4 gcc-4.7.4/config/warnings.m4
--- gcc-4.7.4-pure/config/warnings.m4	2012-08-06 07:34:27.000000000 -0700
+++ gcc-4.7.4/config/warnings.m4	2018-11-19 13:49:12.900409900 -0800
@@ -72,7 +72,7 @@
    [AS_VAR_SET(acx_Pedantic, no)])
 CFLAGS="$save_CFLAGS"])
 AS_IF([test AS_VAR_GET(acx_Pedantic) = yes],
-      [acx_Var="$acx_Var${acx_Var:+ }-pedantic $1"])
+      [acx_Var="$acx_Var${acx_Var:+ }-Wno-pedantic $1"])
 ])
 AS_VAR_POPDEF([acx_Pedantic])dnl
 m4_popdef([acx_Woptions])dnl
diff -Naur gcc-4.7.4-pure/gcc/configure gcc-4.7.4/gcc/configure
--- gcc-4.7.4-pure/gcc/configure	2018-11-16 18:08:57.177822000 -0800
+++ gcc-4.7.4/gcc/configure	2018-11-19 13:49:13.056410200 -0800
@@ -6398,7 +6398,7 @@
 
 loose_warn=
 save_CFLAGS="$CFLAGS"
-for real_option in -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual; do
+for real_option in -W -Wall -Wno-implicit-fallthrough -Wno-cast-function-type -Wno-stringop-truncation -Wno-expansion-to-defined -Wno-narrowing -Wwrite-strings -Wcast-qual; do
   # Do the check with the no- prefix removed since gcc silently
   # accepts any -Wno-* option on purpose
   case $real_option in
@@ -6608,7 +6608,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $acx_cv_prog_cc_pedantic__Wlong_long__Wvariadic_macros__Woverlength_strings" >&5
 $as_echo "$acx_cv_prog_cc_pedantic__Wlong_long__Wvariadic_macros__Woverlength_strings" >&6; }
 if test $acx_cv_prog_cc_pedantic__Wlong_long__Wvariadic_macros__Woverlength_strings = yes; then :
-  strict_warn="$strict_warn${strict_warn:+ }-pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings"
+  strict_warn="$strict_warn${strict_warn:+ }-Wno-pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings"
 fi
 
 fi
diff -Naur gcc-4.7.4-pure/gcc/configure.ac gcc-4.7.4/gcc/configure.ac
--- gcc-4.7.4-pure/gcc/configure.ac	2014-02-12 08:43:47.000000000 -0800
+++ gcc-4.7.4/gcc/configure.ac	2018-11-19 13:49:13.056410200 -0800
@@ -336,7 +336,7 @@
 # So, we only use -pedantic if we can disable those warnings.
 
 ACX_PROG_CC_WARNING_OPTS(
-	m4_quote(m4_do([-W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual])), [loose_warn])
+	m4_quote(m4_do([-W -Wall -Wno-implicit-fallthrough -Wno-cast-function-type -Wno-stringop-truncation -Wno-expansion-to-defined -Wno-narrowing -Wwrite-strings -Wcast-qual])), [loose_warn])
 ACX_PROG_CC_WARNING_OPTS(
 	m4_quote(m4_do([-Wstrict-prototypes -Wmissing-prototypes])),
 	[c_loose_warn])
diff -Naur gcc-4.7.4-pure/gcc/dfp.c gcc-4.7.4/gcc/dfp.c
--- gcc-4.7.4-pure/gcc/dfp.c	2011-05-09 23:05:20.000000000 -0700
+++ gcc-4.7.4/gcc/dfp.c	2018-11-19 13:49:13.087610200 -0800
@@ -26,6 +26,8 @@
 #include "tm_p.h"
 #include "dfp.h"
 
+#pragma GCC diagnostic ignored "-Wstrict-aliasing"
+
 /* The order of the following headers is important for making sure
    decNumber structure is large enough to hold decimal128 digits.  */
 
diff -Naur gcc-4.7.4-pure/gcc/expmed.c gcc-4.7.4/gcc/expmed.c
--- gcc-4.7.4-pure/gcc/expmed.c	2013-02-13 12:26:51.000000000 -0800
+++ gcc-4.7.4/gcc/expmed.c	2018-11-19 13:49:13.118810300 -0800
@@ -21,6 +21,7 @@
 along with GCC; see the file COPYING3.  If not see
 <http://www.gnu.org/licenses/>.  */
 
+#pragma GCC diagnostic ignored "-Warray-bounds"
 
 #include "config.h"
 #include "system.h"
diff -Naur gcc-4.7.4-pure/gcc/gengtype.c gcc-4.7.4/gcc/gengtype.c
--- gcc-4.7.4-pure/gcc/gengtype.c	2011-11-08 08:47:16.000000000 -0800
+++ gcc-4.7.4/gcc/gengtype.c	2018-11-19 13:49:13.150010300 -0800
@@ -1476,7 +1476,8 @@
     "along with GCC; see the file COPYING3.  If not see\n",
     "<http://www.gnu.org/licenses/>.  */\n",
     "\n",
-    "/* This file is machine generated.  Do not edit.  */\n"
+    "/* This file is machine generated.  Do not edit.  */\n",
+    "\n#pragma GCC diagnostic ignored \"-Wswitch-bool\"\n"
   };
   outf_p f;
   size_t i;
diff -Naur gcc-4.7.4-pure/libcpp/configure gcc-4.7.4/libcpp/configure
--- gcc-4.7.4-pure/libcpp/configure	2018-11-16 18:08:57.225828100 -0800
+++ gcc-4.7.4/libcpp/configure	2018-11-19 13:49:13.352810700 -0800
@@ -4688,7 +4688,7 @@
 
 warn=
 save_CFLAGS="$CFLAGS"
-for real_option in -W -Wall -Wno-narrowing -Wwrite-strings \
+for real_option in -W -Wall -Wno-implicit-fallthrough -Wno-cast-function-type -Wno-stringop-truncation -Wno-expansion-to-defined -Wno-narrowing -Wwrite-strings \
 			  -Wmissing-format-attribute; do
   # Do the check with the no- prefix removed since gcc silently
   # accepts any -Wno-* option on purpose
@@ -4811,7 +4811,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $acx_cv_prog_cc_pedantic__Wlong_long" >&5
 $as_echo "$acx_cv_prog_cc_pedantic__Wlong_long" >&6; }
 if test $acx_cv_prog_cc_pedantic__Wlong_long = yes; then :
-  WARN_PEDANTIC="$WARN_PEDANTIC${WARN_PEDANTIC:+ }-pedantic -Wno-long-long"
+  WARN_PEDANTIC="$WARN_PEDANTIC${WARN_PEDANTIC:+ }-Wno-pedantic -Wno-long-long"
 fi
 
 fi
diff -Naur gcc-4.7.4-pure/libcpp/configure.ac gcc-4.7.4/libcpp/configure.ac
--- gcc-4.7.4-pure/libcpp/configure.ac	2013-04-03 08:43:55.000000000 -0700
+++ gcc-4.7.4/libcpp/configure.ac	2018-11-19 13:49:13.352810700 -0800
@@ -33,7 +33,7 @@
 # Figure out what compiler warnings we can enable.
 # See config/warnings.m4 for details.
 
-ACX_PROG_CC_WARNING_OPTS([-W -Wall -Wno-narrowing -Wwrite-strings \
+ACX_PROG_CC_WARNING_OPTS([-W -Wall -Wno-implicit-fallthrough -Wno-cast-function-type -Wno-stringop-truncation -Wno-expansion-to-defined -Wno-narrowing -Wwrite-strings \
 			  -Wmissing-format-attribute], [warn])
 ACX_PROG_CC_WARNING_OPTS([-Wstrict-prototypes -Wmissing-prototypes \
 			  -Wold-style-definition -Wc++-compat], [c_warn])
diff -Naur gcc-4.7.4-pure/libdecnumber/configure gcc-4.7.4/libdecnumber/configure
--- gcc-4.7.4-pure/libdecnumber/configure	2012-08-06 07:34:27.000000000 -0700
+++ gcc-4.7.4/libdecnumber/configure	2018-11-19 13:49:13.384010800 -0800
@@ -3248,7 +3248,7 @@
 
 WARN_CFLAGS=
 save_CFLAGS="$CFLAGS"
-for real_option in -W -Wall -Wwrite-strings -Wstrict-prototypes \
+for real_option in -W -Wall -Wno-implicit-fallthrough -Wno-cast-function-type -Wno-stringop-truncation -Wno-expansion-to-defined -Wwrite-strings -Wstrict-prototypes \
 			  -Wmissing-prototypes -Wold-style-definition \
 			  -Wmissing-format-attribute -Wcast-qual; do
   # Do the check with the no- prefix removed since gcc silently
@@ -3326,7 +3326,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $acx_cv_prog_cc_pedantic__Wlong_long" >&5
 $as_echo "$acx_cv_prog_cc_pedantic__Wlong_long" >&6; }
 if test $acx_cv_prog_cc_pedantic__Wlong_long = yes; then :
-  WARN_PEDANTIC="$WARN_PEDANTIC${WARN_PEDANTIC:+ }-pedantic -Wno-long-long"
+  WARN_PEDANTIC="$WARN_PEDANTIC${WARN_PEDANTIC:+ }-Wno-pedantic -Wno-long-long"
 fi
 
 fi
diff -Naur gcc-4.7.4-pure/libdecnumber/configure.ac gcc-4.7.4/libdecnumber/configure.ac
--- gcc-4.7.4-pure/libdecnumber/configure.ac	2010-09-06 07:52:54.000000000 -0700
+++ gcc-4.7.4/libdecnumber/configure.ac	2018-11-19 13:49:13.384010800 -0800
@@ -37,7 +37,7 @@
 # Figure out what compiler warnings we can enable.
 # See config/warnings.m4 for details.
 
-ACX_PROG_CC_WARNING_OPTS([-W -Wall -Wwrite-strings -Wstrict-prototypes \
+ACX_PROG_CC_WARNING_OPTS([-W -Wall -Wno-implicit-fallthrough -Wno-cast-function-type -Wno-stringop-truncation -Wno-expansion-to-defined -Wwrite-strings -Wstrict-prototypes \
 			  -Wmissing-prototypes -Wold-style-definition \
 			  -Wmissing-format-attribute -Wcast-qual])
 ACX_PROG_CC_WARNING_ALMOST_PEDANTIC([-Wno-long-long])
diff -Naur gcc-4.7.4-pure/libiberty/configure gcc-4.7.4/libiberty/configure
--- gcc-4.7.4-pure/libiberty/configure	2012-01-22 22:25:28.000000000 -0800
+++ gcc-4.7.4/libiberty/configure	2018-11-19 13:49:13.415210800 -0800
@@ -3920,7 +3920,7 @@
 
 ac_libiberty_warn_cflags=
 save_CFLAGS="$CFLAGS"
-for real_option in -W -Wall -Wwrite-strings -Wc++-compat \
+for real_option in -W -Wall -Wno-implicit-fallthrough -Wno-cast-function-type -Wno-stringop-truncation -Wno-expansion-to-defined -Wwrite-strings -Wc++-compat \
 			  -Wstrict-prototypes; do
   # Do the check with the no- prefix removed since gcc silently
   # accepts any -Wno-* option on purpose
@@ -3996,7 +3996,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $acx_cv_prog_cc_pedantic_" >&5
 $as_echo "$acx_cv_prog_cc_pedantic_" >&6; }
 if test $acx_cv_prog_cc_pedantic_ = yes; then :
-  ac_libiberty_warn_cflags="$ac_libiberty_warn_cflags${ac_libiberty_warn_cflags:+ }-pedantic "
+  ac_libiberty_warn_cflags="$ac_libiberty_warn_cflags${ac_libiberty_warn_cflags:+ }-Wno-pedantic "
 fi
 
 fi
diff -Naur gcc-4.7.4-pure/libiberty/configure.ac gcc-4.7.4/libiberty/configure.ac
--- gcc-4.7.4-pure/libiberty/configure.ac	2011-08-22 09:54:02.000000000 -0700
+++ gcc-4.7.4/libiberty/configure.ac	2018-11-19 13:49:13.415210800 -0800
@@ -133,7 +133,7 @@
 AC_SYS_LARGEFILE
 AC_PROG_CPP_WERROR
 
-ACX_PROG_CC_WARNING_OPTS([-W -Wall -Wwrite-strings -Wc++-compat \
+ACX_PROG_CC_WARNING_OPTS([-W -Wall -Wno-implicit-fallthrough -Wno-cast-function-type -Wno-stringop-truncation -Wno-expansion-to-defined -Wwrite-strings -Wc++-compat \
 			  -Wstrict-prototypes], [ac_libiberty_warn_cflags])
 ACX_PROG_CC_WARNING_ALMOST_PEDANTIC([], [ac_libiberty_warn_cflags])
 
